arc_design_contest/2018/NCKU_Vending_Machine/embarc_osp-master/board/board.c

```diff
@@ -0,0 +1,390 @@
+			//vTaskResume(task_handle_wifi);
+#else 
+			if (communication_task) communication_task (NULL);
+			flag = 0;
+#endif
+#else
+			if (task_handle_main) vTaskResume(task_handle_main);
+#endif
+#else
+			EMBARC_PRINTF("Please run NT-Shell command(main) to start your application.\r\n");
+			EMBARC_PRINTF("main command may required some arguments, please refer to example's document.\r\n");
+#endif
+			/* consider to generate a event to notify network is ready */
+		} else if ((flag == 1) && !lwip_pmwifi_isup()){// Wifi Lost connection
+			if(task_handle_communication) vTaskSuspend(task_handle_communication);
+			flag = 0;
+		}
+		vTaskDelay(TASK_WIFI_PERIOD);
+	}
+}
+#endif
+
+static void task_main(void *par)
+{
+	int ercd;
+	EMBARC_PRINTF("Enter to main function....\r\n");
+#if defined(OS_FREERTOS) && defined(MID_LWIP) && !defined(MID_NTSHELL) && !defined(MID_VM)
+	EMBARC_PRINTF("Wait until WiFi connected...\r\n");
+	vTaskSuspend(NULL);
+#endif
+
+	if ((par == NULL) || (((int)par) & 0x3)) {
+	/* null or aligned not to 4 bytes */
+		ercd = _arc_goto_main(0, NULL);
+	} else {
+		MAIN_ARGS *main_arg = (MAIN_ARGS *)par;
+		ercd = _arc_goto_main(main_arg->argc, main_arg->argv);
+	}
+
+#if defined(OS_FREERTOS)
+	EMBARC_PRINTF("Exit from main function, error code:%d....\r\n", ercd);
+	while (1) {
+		vTaskSuspend(NULL);
+	}
+#else
+	while (1);
+#endif
+}
+
+void board_main(void)
+{
+/* board level hardware init */
+	board_init();
+
+// spi swmaphore: declare lock for spi ifdef FREERTOS
+#ifdef OS_FREERTOS
+#ifdef MID_VM
+    spi_mutex = xSemaphoreCreateBinary();
+	xSemaphoreGive(spi_mutex);
+    vm_queue_mutex = xSemaphoreCreateBinary();
+	xSemaphoreGive(vm_queue_mutex);
+#endif
+#endif
+
+/* board level middlware init */
+
+#ifdef MID_COMMON
+	xprintf_setup();
+#endif
+
+#ifdef MID_FATFS
+	if(f_mount(&sd_card_fs, "", 0) != FR_OK) {
+		EMBARC_PRINTF("FatFS failed to initialize!\r\n");
+	} else {
+		EMBARC_PRINTF("FatFS initialized successfully!\r\n");
+	}
+#endif
+
+#ifdef ENABLE_OS
+	os_hal_exc_init();
+#endif
+
+// try if the makefile is normal
+#ifdef MID_VM
+#ifdef OS_FREERTOS
+
+    /* create common-use queue for inter-communication */ 
+    xVMQueue = xQueueCreate(48, sizeof(vm_data));
+
+    /* create 5 queues for specific usage and device */
+	xMainQueue             = xQueueCreate(12, sizeof(vm_data));
+    xOledQueue             = xQueueCreate(12, sizeof(vm_data));
+    xCommunicationQueue    = xQueueCreate(12, sizeof(vm_data));
+    xTempQueue             = xQueueCreate(12, sizeof(vm_data));
+    xNumPadQueue           = xQueueCreate(12, sizeof(vm_data));
+    xDCmotorQueue          = xQueueCreate(12, sizeof(vm_data));
+
+    /* create 5 tasks */
+	xTaskCreate((TaskFunction_t)vm_task,            "vending machine",    TASK_STACK_SIZE_VM,            NULL, TASK_PRI_VM,            &task_handle_vm);
+	xTaskCreate((TaskFunction_t)LED_task,           "led blinky",         TASK_STACK_SIZE_LED,           NULL, TASK_PRI_LED,           &task_handle_led);
```

Comment:
Can not accept modification to board.c to create user taskes. 
@fanghuaqi This PR is also unable to pass CI, please take a look. THX.

Reply:
Yeah, @cstandy you need to create tasks in your application code not in embARC OSP code, and could you make a list of the changes in diff or send a pull request to embarc_osp, that you have made  to embARC OSP, let us see what can be modified and added in application code not osp code.